### PR TITLE
Add mypy to ansible-core tests.

### PR DESCRIPTION
**Pull request by mattclay**
##### SUMMARY

This is a proof-of-concept demo for adding `mypy` to the sanity tests for ansible-core only (not collections).

This should help with ongoing discussions the core team is having regarding use of automated type checking.

##### ISSUE TYPE

Feature Pull Request

##### COMPONENT NAME

ansible-test

### CODE REVIEW

**Comment by webknjaz** [comment]
@mattclay are you planning to replace my PR with this or is it a companion change? I haven't had a chance to look closely but it seems you're attempting to pass individual modules to mypy which is rather wrong: it will confuse multiple contexts and may even fail early when the module names are the same because when you don't pass the package context, it does not have a way to distinguish between the importables.

**Comment by mattclay** [comment]
@webknjaz This implementation would replace the one in your PR. Can you provide a working example of the issue(s) with passing module names to mypy? I didn't find anything in the documentation or testing that indicate doing so is problematic.

**Comment by webknjaz** [comment]
> Can you provide a working example of the issue(s) with passing module names to mypy? I didn't find anything in the documentation or testing that indicate doing so is problematic.

I think it's partially described in https://github.com/python/mypy/issues/10428.

Also, here's what you can hit in this repo:
```console
$ mypy test/integration/targets/pip/files/setup.py setup.py
setup.py: error: Duplicate module named "setup" (also at "test/integration/targets/pip/files/setup.py")
setup.py: note: Are you missing an __init__.py? Alternatively, consider using --exclude to avoid checking one of them.
Found 1 error in 1 file (errors prevented further checking)
```
or
```console
$ mypy --namespace-packages test/ lib/ansible
test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/modules/bad.py: error: Duplicate module named "bad" (also at "test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/lookup/bad.py")
test/integration/targets/ansible-test/ansible_collections/ns/col/plugins/modules/bad.py: note: Are you missing an __init__.py? Alternatively, consider using --exclude to avoid checking one of them.
Found 1 error in 1 file (errors prevented further checking)
```
(a subset of instances of this specific error can be fixed by `--namespace-packages`)

Although, it's not exactly the corner case I've had in my mind.

**Comment by webknjaz** [comment]
Another case I was thinking about was this. Imagine you've got two related modules:
```console
$ head mod*.py               
==> mod1.py <==
from mod2 import func

func(1)

==> mod2.py <==
def func(x: int) -> None:
    ...
```

And you run mypy on them, and it's fine:
```console
$ mypy --strict --no-incremental mod2.py mod1.py 
Success: no issues found in 2 source files

$ mypy --strict --no-incremental mod1.py
Success: no issues found in 1 source file

$ mypy --strict --no-incremental mod2.py
Success: no issues found in 1 source file
```

Then, you change `mod2.py`:
```console
$ cat mod2.py                                
def func(x: str) -> None:
    ...
```
The CI sees that it's changed and lints it:
```
$ mypy --strict --no-incremental mod2.py        
Success: no issues found in 1 source file
```

The PR is green because the changed file does not have problems within itself so you merge it.
And then, whatever new PR tries to change `mod1.py`, it'll hit an unexpected unrelated error:
```console
$ mypy --strict --no-incremental mod1.py 
mod1.py:3: error: Argument 1 to "func" has incompatible type "int"; expected "str"
Found 1 error in 1 file (checked 1 source file)
```

I think this is why such a setup is prone to cause confusion — when you pass individual modules packages, it doesn't go beyond what's passed and doesn't check how it affects the rest of the project which is rather unexpected.

File: test/lib/ansible_test/_internal/commands/sanity/mypy.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            # They are primarily useful for listing individual 'ignore_missing_imports' entries instead of using a global ignore.
+            '--config-file', config_path,
+        ]
+
+        cmd.extend(context_paths)
+
+        try:
+            stdout, stderr = intercept_python(args, virtualenv_python, cmd, env, capture=True)
+
+            if stdout or stderr:
```
Comment (webknjaz) [inline test/lib/ansible_test/_internal/commands/sanity/mypy.py:219]: I feel like this should be moved to `else:`

File: test/lib/ansible_test/_internal/commands/sanity/mypy.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    """Sanity test which executes mypy."""
+    ansible_only = True
+
+    def filter_targets(self, targets):  # type: (t.List[TestTarget]) -> t.List[TestTarget]
+        """Return the given list of test targets, filtered to include only those relevant for the test."""
+        return [target for target in targets if os.path.splitext(target.path)[1] == '.py' and (
+                target.path.startswith('lib/ansible/') or target.path.startswith('test/lib/ansible_test/_internal/'))]
+
+    @property
+    def error_code(self):  # type: () -> t.Optional[str]
```
Comment (webknjaz) [inline test/lib/ansible_test/_internal/commands/sanity/mypy.py:69]: Looks like you mix type comments with type annotations (at the very bottom) in this module — why not use a consistent/native style?

File: lib/ansible/module_utils/distro/_distro.py [hunk 3]
```diff
@@ -41,6 +41,39 @@
 
 __version__ = "1.6.0"
 
+# Use `if False` to avoid an ImportError on Python 2. After dropping Python 2
+# support, can use typing.TYPE_CHECKING instead. See:
+# https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING
+if False:  # pragma: nocover
```
Comment (webknjaz) [inline lib/ansible/module_utils/distro/_distro.py]: Consider using the idiom I used in the depresolver code: https://github.com/ansible/ansible/blob/aa4f23d/lib/ansible/galaxy/dependency_resolution/dataclasses.py#L16-L21.

We could even put that try/except into some compat module that could be reused.

**Comment by mattclay** [comment]
@webknjaz Thanks for the examples. However, they don't appear to be applicable to the changes in this PR.

The first and second example both use files which aren't tested by `ansible-test` with `mypy`. The second example uses an option that `ansible-test` doesn't use. I haven't found a way to trigger the duplicate module behavior using `ansible-test`.

The third case is a general issue with any of our tests that involve dependencies between files, and is not specific to `mypy`. The change detection in `ansible-test` takes these dependencies into account when running tests to avoid the issue.

File: lib/ansible/module_utils/distro/_distro.py (hunk 3 above)
Comment (mattclay) [reply lib/ansible/module_utils/distro/_distro.py]: This code comes from the original `distro` implementation, so it shouldn't be changed unless it's broken.

File: test/lib/ansible_test/_internal/commands/sanity/mypy.py (hunk 2 above)
Comment (mattclay) [reply test/lib/ansible_test/_internal/commands/sanity/mypy.py:69]: Nearly all of the type annotations in `ansible-test` are comment based. The main exception is for dataclasses, which require native type annotations.

File: lib/ansible/module_utils/distro/_distro.py (hunk 3 above)
Comment (webknjaz) [reply lib/ansible/module_utils/distro/_distro.py]: In that case, maybe submit this separately to avoid confusion?

**Pull request merged**